### PR TITLE
v5 extension detection

### PR DESCRIPTION
**Pull request by jnsdls**
## PR-Codex overview
This PR adds new promise conditionals for `allOf` and `oneOf`, updates extension methods, and introduces a new package `@shazow/whatsabi`.

### Detailed summary
- Added `allOf` and `oneOf` promise conditionals
- Updated extension methods for ERC20 contracts
- Introduced `@shazow/whatsabi` package
- Updated bytecode resolution and function selector retrieval

> The following files were skipped due to too many changes: `apps/dashboard/src/contract-ui/tabs/tokens/page.tsx`, `apps/dashboard/src/pages/[chain_id]/[...paths].tsx`, `packages/thirdweb/src/extensions/erc20/read/isERC20.ts`, `packages/thirdweb/src/extensions/erc20/drops/read/getActiveClaimCondition.ts`, `packages/thirdweb/src/utils/promise/conditionals.test.ts`, `packages/thirdweb/src/extensions/erc20/write/mintTo.test.ts`, `packages/thirdweb/src/exports/extensions/common.ts`, `apps/dashboard/src/contract-ui/tabs/settings/page.tsx`, `apps/dashboard/src/contract-ui/hooks/useRouteConfig.tsx`, `pnpm-lock.yaml`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Comment by jnsdls** [comment]
* **#4260** <a href="https://app.graphite.dev/github/pr/thirdweb-dev/js/4260?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @jnsdls and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/thirdweb/src/extensions/erc20/read/isERC20.ts [hunk 1]
```diff
@@ -13,8 +22,20 @@ import { supportsInterface } from "../../erc165/__generated__/IERC165/read/suppo
...
+  // see: https://github.com/OpenZeppelin/openzeppelin-contracts/issues/3575
+  // see: https://ethereum.org/en/developers/docs/standards/tokens/erc-20/
+
+  return allOf(
+    isNameSupported(options.contract),
+    isSymbolSupported(options.contract),
+    isDecimalsSupported(options.contract),
+    isTotalSupplySupported(options.contract),
+    isBalanceOfSupported(options.contract),
+    isBalanceOfSupported(options.contract),
```
Comment (gregfromstl) [inline packages/thirdweb/src/extensions/erc20/read/isERC20.ts:35]: balanceOf twice

Comment (jnsdls) [reply packages/thirdweb/src/extensions/erc20/read/isERC20.ts:35]: good catch!

File: packages/thirdweb/package.json [hunk 2]
```diff
@@ -162,6 +162,7 @@
     "@radix-ui/react-focus-scope": "1.1.0",
     "@radix-ui/react-icons": "1.3.0",
     "@radix-ui/react-tooltip": "1.1.2",
+    "@shazow/whatsabi": "^0.14.1",
```
Comment (joaquim-verges) [inline packages/thirdweb/package.json:165]: ugh bundle size gets a big hit... not sure if its worth adding as a hard dep for a mostly dashboard specific use case.

did you look into extracting just the bytecode fn extraction bit?

File: packages/thirdweb/src/extensions/erc20/drops/read/getActiveClaimCondition.ts [hunk 3]
```diff
@@ -1,8 +1,10 @@
+import type { ThirdwebContract } from "../../../../contract/contract.js";
 import type { BaseTransactionOptions } from "../../../../transaction/types.js";
 import type { ClaimCondition } from "../../../../utils/extensions/drops/types.js";
-import { claimCondition } from "../../__generated__/DropSinglePhase/read/claimCondition.js";
-import { getActiveClaimConditionId } from "../../__generated__/IDropERC20/read/getActiveClaimConditionId.js";
-import { getClaimConditionById } from "../../__generated__/IDropERC20/read/getClaimConditionById.js";
+import { allOf, oneOf } from "../../../../utils/promise/conditionals.js";
+import * as Single from "../../__generated__/DropSinglePhase/read/claimCondition.js";
+import * as MultiActiveId from "../../__generated__/IDropERC20/read/getActiveClaimConditionId.js";
+import * as MultiById from "../../__generated__/IDropERC20/read/getClaimConditionById.js";
```
Comment (joaquim-verges) [inline packages/thirdweb/src/extensions/erc20/drops/read/getActiveClaimCondition.ts:7]: we doin this now huh

File: packages/thirdweb/src/extensions/erc20/read/isERC20.ts [hunk 4]
```diff
@@ -13,8 +22,19 @@
...
 export function isERC20(options: BaseTransactionOptions) {
-  return supportsInterface({
-    contract: options.contract,
-    interfaceId: "0x36372b07",
-  });
+  // there is no trustworthy way to check if a contract is ERC20 via ERC165, so we do this manually.
+  // see: https://github.com/OpenZeppelin/openzeppelin-contracts/issues/3575
+  // see: https://ethereum.org/en/developers/docs/standards/tokens/erc-20/
+
+  return allOf(
```
Comment (joaquim-verges) [inline packages/thirdweb/src/extensions/erc20/read/isERC20.ts:29]: feels like this should be the fallback if supportsInterface fails?  would still try to rely on it, its pretty widespread i think

File: packages/thirdweb/src/utils/bytecode/selectors.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import { selectorsFromBytecode } from "@shazow/whatsabi";
+import type { ThirdwebContract } from "../../contract/contract.js";
+import { resolveImplementation } from "../../utils/bytecode/resolveImplementation.js";
+import { LruMap } from "../../utils/caching/lru.js";
+
+const SelectorCache = new LruMap<string[]>(8192);
```
Comment (joaquim-verges) [inline packages/thirdweb/src/utils/bytecode/selectors.ts:6]: 8k contracts in cache? seems like a lot no?

File: packages/thirdweb/src/utils/promise/conditionals.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+/**
+ * Checks if all promises in the given array resolve to a truthy value.
+ *
+ * @param promises - Promises to check.
+ * @returns A promise that resolves to `true` if all promises resolve to a truthy value, otherwise resolves to `false`.
+ * @internal
+ */
+export async function allOf(...promises: Promise<boolean>[]): Promise<boolean> {
```
Comment (joaquim-verges) [inline packages/thirdweb/src/utils/promise/conditionals.ts:30]: thats cute

File: packages/thirdweb/src/utils/bytecode/selectors.ts (hunk 5 above)
Comment (jnsdls) [reply packages/thirdweb/src/utils/bytecode/selectors.ts:6]: it's really just "don't get TOO big" - the other option was just a normal map, there's no reason to ever do it again once we've done it once

File: packages/thirdweb/src/extensions/erc20/read/isERC20.ts (hunk 4 above)
Comment (jnsdls) [reply packages/thirdweb/src/extensions/erc20/read/isERC20.ts:29]: not for ERC20 (see the comments above)

File: packages/thirdweb/src/extensions/erc20/drops/read/getActiveClaimCondition.ts (hunk 3 above)
Comment (jnsdls) [reply packages/thirdweb/src/extensions/erc20/drops/read/getActiveClaimCondition.ts:7]: it seems to work fine :)

File: packages/thirdweb/package.json (hunk 2 above)
Comment (jnsdls) [reply packages/thirdweb/package.json:165]: it doesn't actually make a big difference as long as ur not using the detection (size limit doesn't fail even) - I extracted it out first but we need _most_ of the logic and the rest is correctly tree-shaken

**Pull request closed without merging**